src/PostgREST/Request/ApiRequest.hs

```diff
@@ -494,7 +494,21 @@ findProc qi argumentsKeys paramsAsSingleObject allProcs contentType isInvPost =
     [proc] -> Right proc
     procs  -> Left $ AmbiguousRpc (toList procs)
   where
-    matchProc = filter matchesParams $ M.lookupDefault mempty qi allProcs -- first find the proc by name
+    matchProc =
+      if null (fst findProcPartition) && isInvPost && contentType == CTApplicationJSON
+        then snd findProcPartition -- use the fallback function if none is found with the provided parameters
+        else fst findProcPartition
+    findProcPartition = overloadedProcPartition $ M.lookupDefault mempty qi allProcs -- first find the proc by name
+    -- The partition obtained has the form (overloadedProcs,fallbackProcs)
+    -- where fallbackProcs are functions with a single unnamed parameter of type json or jsonb
+    overloadedProcPartition procs = foldr select ([],[]) procs
+    select proc ~(ts,fs)
+      | matchesParams proc             = (proc:ts,fs)
+      | hasSingleUnnamedJsonParam proc = (ts,proc:fs)
+      | otherwise                      = (ts,fs)
+    hasSingleUnnamedJsonParam proc = case pdParams proc of
+      [ProcParam ppName ppType _ _] -> T.null ppName && ppType `elem` ["json", "jsonb"]
+      _ -> False
```

Comment:
Should we maybe put this bit on the `ProcDescription` type at schema cache creation time already?

We currently have:

```haskell
data ProcDescription = ProcDescription
[...]
  , pdParams      :: [ProcParam]
[...]
```

and could change it to something like

```haskell
data ProcDescription = ProcDescription
[...]
  , pdUnnamedParam     :: Maybe ProcUnnamedParamType
  , pdNamedParams      :: [ProcParam]
[...]

data ProcUnnamedParamType = SingleBytea | SingleJson | SingleText
```

And then filter the procs at schema cache creation time already:
- remove all procs with more than 1 unnamed param
- remove all procs with 1 unnamed param + named params (side note: should we add a test-case for such a situation?)
- remove all unnamed args from `pdNamedParams`

I think this could make the whole logic here (and below) a bit more readable?

Reply:
I like the idea of doing the filter at schema cache creation time, but separating the parameters, while it does make the logic more readable here, it complicates it a bit on other parts, like [here](https://github.com/PostgREST/postgrest/blob/main/src/PostgREST/Request/DbRequestBuilder.hs#L360) or in the OpenApi for instance. 

Maybe adding the fallback type and using it as the filter if it exists could be a bit simpler?

```haskell
data ProcDescription = ProcDescription
[...]
  , pdFallbackType :: Maybe ProcFallbackType
[...]

data ProcFallbackType = Bytea | Json | Text
```

> - remove all procs with more than 1 unnamed param
> - remove all procs with 1 unnamed param + named params (side note: should we add a test-case for such a situation?)

Removing all procs would make functions like this `create or replace function api.test_overload(c int, json default '{}')` to not work causing a breaking change, although I can't think of a use case right now. A solution would be to remove those where at least one unnamed parameter doesn't have a default value.

Reply:
Ok, replace "unnamed param" with "unnamed required param" then, in my list above.

> Maybe adding the fallback type and using it as the filter if it exists could be a bit simpler?

Ok, let's do it like that. Let's call it like this:

```haskell
data ProcDescription = ProcDescription
[...]
  , pdPositionalParam :: Maybe ProcPositionalParam
[...]

data ProcPositionalParam = SingleBytea | SingleJson | SingleText
```

I think we need the `Single` prefix, otherwise we'd get conflicts with built-in `Text` very quickly.

Positional is better than "Fallback", because the first is descriptive of what is in the database, while the second is already hinting at how it might be used. But `ProcDescription` is really just that - a description.

Reply:
With the latest changes, I think it's actually quite readable already. Maybe we don't need to change the typing at all for now.

Reply:
Yes, I agree, it looks good right now. I'll leave this part as is, fix some tests and then this should be ready for a last review and merge.